kconfig: fix dt_node_has_prop

barnas-michal opened the pull request:
The Kconfig function "dt_node_has_prop" was using label as its
parameter, where other functions use either chosen or path.
The documentation says that the parameter is path, so this patch
makes the function as documentation says and as other functions
in the file.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
kconfig: fix dt_node_has_prop and add nodelabel functions

The Kconfig function "dt_node_has_prop" was using label as its
parameter, where other functions use either chosen or path.
The documentation says that the parameter is path, so this patch
makes the function as documentation says and as other functions
in the file.
The additional nodelabel functions were added as counterparts that
are using nodes labels instead of paths.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

erwango reviewed (changes requested):
Using label instead of path was implemented as purpose, as this is more convenient.
Here is an example of a symbol defined thanks a function using path:
```
config FLASH_SIZE
	default $(dt_node_reg_size_int,/soc/flexram@40028000/itcm@0,0,K) if SOC_SERIES_IMX_RT11XX
	default $(dt_node_reg_size_int,/soc/flexram@400b0000/itcm@0,0,K) if SOC_SERIES_IMX_RT10XX
```
We can see that using path:
-is not quite easy to read
-prevents abstraction

If this function was implemented using label, we could have:
```
config FLASH_SIZE
	default $(dt_node_reg_size_int,itcm,0,K)
```
Note that thanks to dtc checks, we're sure that label is unique and can be used safely.
Forcing to use path instead of label is the contrary of an improvement.

So, rather than fixing the function, I suggest to fix the doc.
And if you have time to spend, then convert the path using functions.

Also, please don't change a function implementation if you don't update the users.

erwango commented:
> > Using label instead of path was implemented as purpose, as this is more convenient.
> 
> I understand that, but it's not consistent with rest of the functions in the file. I would rather make this function as the rest, to operate on the paths, and make another functions that operate on labels but with name that suggests the difference (x_by_label).

Ok, then please do.

> > Note that thanks to dtc checks, we're sure that label is unique and can be used safely. Forcing to use path instead of label is the contrary of an improvement.
> 
> The functions that are using the paths, can be provided with aliases, since the devicetree lib checks if path starts with "/", if not, it looks for aliases. This will also guarantee the uniqueness.

If labels, which are there anyway, are sufficient to provide what we need, I don't think this is a wise choice to add aliases just to satisfy the requirements of poorly designed kconfig functions. We should aim at simplicity whenever it is possible.

> > Also, please don't change a function implementation if you don't update the users.
> 
> How is this done? In changelog or something?

I mean in tree users at least, for which grepping the code should be sufficient.
Then, I'm not sure if this API is exposed to out of tree users, but in which case, a deprecation message or warning should be considered. For this specific case of Kconfig API, I'm not sure how to proceed though.

barnas-michal committed:
docs: add missing Kconfig functions

Add missing Kconfig functions to documentation about Kconfig
preprocessor functions.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

semihalf-barnas-michal commented:
> Ok, then please do.

Done

> If labels, which are there anyway, are sufficient to provide what we need, I don't think this is a wise choice to add aliases just to satisfy the requirements of poorly designed kconfig functions. We should aim at simplicity whenever it is possible.

I agree with you.

> I mean in tree users at least, for which grepping the code should be sufficient. Then, I'm not sure if this API is exposed to out of tree users, but in which case, a deprecation message or warning should be considered. For this specific case of Kconfig API, I'm not sure how to proceed though.

Right, changed the usage to nodelabel function.

erwango reviewed (changes requested):
Thanks for this work and taking my remarks into account.
2 minor comments, otherwise LGTM.

semihalf-barnas-michal commented:
> Thanks for this work and taking my remarks into account. 2 minor comments, otherwise LGTM.

Fixed, thanks for your review :)

mbolivar-nordic commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -337,19 +333,38 @@ def dt_node_has_bool_prop(kconf, _, path, prop):
+
+    return _dt_node_has_bool_prop_generic(edt.get_node, path, prop)
 
+def dt_nodelabel_has_bool_prop(kconf, _, label, prop):
+    """
+    This function takes a 'label' and looks for an EDT node with that label.
+    If it finds an EDT node, it will look to see if that node has a boolean
+    property by the name of 'prop'.  If the 'prop' exists it will return "y"
+    otherwise we return "n".
+    """
```
Functions or API calls in the style of "has bool prop" are usually the wrong way to go.

For example, sometimes people will use [DT_NODE_HAS_PROP(foo, bar)](https://docs.zephyrproject.org/latest/reference/devicetree/api.html#group__devicetree-generic-exist_1gacce67bf20541cd2d07d8540058964692) to test a boolean property, but this should be [DT_PROP(foo, bar)](https://docs.zephyrproject.org/latest/reference/devicetree/api.html#group__devicetree-generic-prop_1ga8e1fd9ebacd85d2013df027d041d506b) instead. DT_PROP knows that the property is boolean because of the binding. Therefore, presence is true, and absence is false.

I have the same objection to this function here; this is semantically wrong because "having" a boolean property is the same thing as the property being true. So I'm not sure this is needed. This is different than dt_nodelabel_has_prop because some properties are optional. A boolean property is always optional, but its presence or absence have different semantics.

I would advocate for a `dt_nodelabel_bool_prop` function instead, which either returns y or n for boolean properties (and returns n by default). Then you can later on add `dt_nodelabel_int_prop_int`, `dt_nodelabel_int_prop_hex`, etc., and the naming will all be consistent.

semihalf-barnas-michal replied:
It sounds reasonable. You're right about the checking for boolean properties, I've used wrong functions too in the past.
Also the naming is more consistent with the int ones. Thanks, done that!
